runtime/vm/resolvesupport.cpp

```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	J9Class *throwable = internalFindKnownClass(vmThread, J9VMCONSTANTPOOL_JAVALANGTHROWABLE, J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
+	UDATA resolved = 0;
+
+retry:
+	/* Check if already resolved */
+	if (NULL != value) {
+		resolved = 1;
+	} else if (NULL != ramCPEntry->exception) {
+		if (isSameOrSuperClassOf(throwable, J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, ramCPEntry->exception))) {
+			/* Resolved with exception */
+			VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+			resolved = 1;
+		} else if (ramCPEntry->exception == vmThread->javaVM->voidReflectClass->classObject) {
+			/* Resolved with null reference */
+			resolved = 1;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			// yield(vmThread);
+			goto retry;
+		}
+	}
+
+	/* Enter if not previously resolved */
+	if (0 == resolved) {
+		/* Set exception to current threadObject as a mutex to prevent other threads from entering */
+		if ((ramCPEntry->exception != vmThread->threadObject) &&
+			(0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, ramClass, &ramCPEntry->exception, NULL, vmThread->threadObject))
+		) {
+			goto retry;
+		}
+
+		/* Walk bsmData */
+		for (U_32 i = 0; i < bsmIndex; i++) {
+			bsmData += bsmData[1] + 2;
+		}
+
+		/* Invoke BootStrap method*/
+		sendResolveConstantDynamic(vmThread, ramCP, cpIndex, nameAndSig, bsmData);
+		value = (j9object_t)vmThread->returnValue;
+
+		j9object_t exceptionObject = NULL;
+		if (NULL != vmThread->currentException) {
+			/* Allocate the exception object in tenure */
+			exceptionObject = gcFuncs->j9gc_objaccess_asConstantPoolObject(
+											vmThread,
+											vmThread->currentException,
+											J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
+			if (NULL == exceptionObject) {
+				setHeapOutOfMemoryError(vmThread);
+				exceptionObject = vmThread->outOfMemoryError;
+			}
+		} else if (NULL == value) {
+			/* Java.lang.Object is used as a special flag to indicate null reference */
+			exceptionObject = vmThread->javaVM->voidReflectClass->classObject;
+		} else {
+			value = gcFuncs->j9gc_objaccess_asConstantPoolObject(
+									vmThread,
+									value,
+									J9_GC_ALLOCATE_OBJECT_TENURED | J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE | J9_GC_ALLOCATE_OBJECT_HASHED);
+
+			if (NULL == value) {
+				setHeapOutOfMemoryError(vmThread);
+				exceptionObject = vmThread->outOfMemoryError;
+			} else if (ramCPEntry->exception != vmThread->threadObject) {
+				goto retry;
+			} else {
+				ramCPEntry->value = value;
+			}
+		}
+
+		if (ramCPEntry->exception == vmThread->threadObject) {
+			ramCPEntry->exception = exceptionObject;
```

Comment:
might need a write barrier before this. I'm not sure how much reordering the compiler will do, but if it is possible to reoder this before the write to `ramCPEntry->value`, then you definitely need a write barrier.